api.py

```diff
@@ -0,0 +1,75 @@
+from functools import lru_cache, partial
+from typing import Callable, Dict, List, Optional, Tuple
+import os
+
+import flask
+
+app = flask.Flask(__name__)
+
+FILESYSTEM_ROOT = os.path.abspath(os.sep)
+DATASETS_ROOT = os.path.join(FILESYSTEM_ROOT, 'data')
+
+
+def readable(dataset_name: str) -> str:
+    """Convert name on disk to readable for the user"""
+    return dataset_name.replace('_', ' ')
+
+
+def get_datasets() -> List[Dict[str, str]]:
+    """"Get datasets available in the store"""
+    datasets = filter(os.path.isdir, os.listdir(DATASETS_ROOT))
+    return [{"name": readable(name), "value": name} for name in datasets]
+
+
+def substitute_tags(tag_map: Dict[str, str], text: str) -> str:
+    """Substitute tags from the text for corresponding values in the map"""
+    for tag in tag_map:
+        text = text.replace('"' + tag + '"', tag_map[tag])
+    return text
+
+
+Substitutor = Optional[Callable[[str], str]]
+
+
+def datasets_substitutor() -> Substitutor:
+    """Factory of datasets substitutor"""
+    datasets = get_datasets()
+    parsed = str(datasets).replace("'", '"')
+    return partial(substitute_tags, {'$DATASETS': parsed})
+
+
+SubstitutorFactory = Callable[[], Substitutor]
+Response = Tuple[str, int]
+NOT_FOUND = "", 404
+
+
+def file_from_disk(substitutor_factory: SubstitutorFactory, path: str) -> Response:
```

Comment:
Just curious - why are you introducing here SubstitutionFactory? Why just not pass a Substitutor object? It looks redundant to me but maybe there is something deeper in this.

Reply:
I had `Substitutor` at first, but `Substitutor` object has already all datasets inside, while we must enforce checking for available datasets for each call once again. Therefore, if datasets are obtained during creation of `Substitutor`, we've just delayed creation until this function is called.